File: Project2_MultivariatePricing/Run_final_project_group2A.m
% Final FE Project: Multivariate Levy Pricing
% Group 2A

% Description:
% The project focus on the comparison of 2 different methods of model
% pricing; an exponential Levy framework with multiple parameters and a
% Black framework which takes consideration of the volatility only.
% Moreover, the project requires the pricing of an exotic derivative
% between the S&P500 and the EURO50 markets.

% Maspes Marco
% Tarditi Andrea
% Torba Matteo

%% Clearing the workspace
clear all; close all; clc;
warning('off', 'all');

% Start run time computation
tic;

% Fix random seed for the simulation
rng(42);

%% Flags

% flag = 1 for plotting enabled
% flag = 0 for plotting disabled

flag = 0;

% save_results = 1 for text saving enabled
% save_results = 0 for text saving disabled

save_results = 0;

%% Load folders
% Introduced multiple folders for a better division of the scripts

addpath('data');
addpath('forward price');
addpath('calibration');
addpath('general');
addpath('pricing_certificate');
addpath('plotting');
addpath('tables_multicalib');

%% Loading of the matrices
% Loading of the initial matrices and structures

data_USA = load("OptionData.mat").mkt;
data_EU = load("OptionData.mat").mkt_EU;

SP500_EUR50 = load("SPXSX5Ereturns.mat").Returns;

%% Initialization of the base parameters

% Settlement date:
date_settlement = datenum('2023-07-09');

% Dates Vector:
dates_EU = datenum(data_EU.datesExpiry);
dates_USA = datenum(data_USA.datesExpiry);

% Computation of the historical correlation
rho_historical = hist_corr(SP500_EUR50);

%% Year frac conventions

conv_ACT360 = 2; conv_ACT365 = 3; conv_30360_EU = 6;

%% Plot of initial options

if flag == 1
    dataset_exploration(data_EU, data_USA)
end

%% Plot of the returns

if flag == 1
    plot_returns(SP500_EUR50,date_settlement)
end

%%

%% POINT 5: Forward Prices
% Computation of the forward prices through the Synthetic Forwards
% procedure

[F0_EU, B_EU] = forward_prices(data_EU, flag);
[F0_USA, B_USA] = forward_prices(data_USA, flag);

if flag == 1
    plot_forwards_discounts(data_EU, data_USA, B_EU, B_USA, F0_EU, F0_USA, date_settlement);
end

%%

%% POINT 6: Calibration

%% Options selection

% Choice of only OTM options for the further calibration
data_EU_OTM = OTM_preprocessing(data_EU, F0_EU);
data_USA_OTM = OTM_preprocessing(data_USA, F0_USA);

% Computing the Delta of Black & Scholes over the OTM Call/Put in order to remove from dataset the OTM options
data_calib_EU = dataset_preprocessing(data_EU_OTM, F0_EU, B_EU, date_settlement, flag);
data_calib_USA = dataset_preprocessing(data_USA_OTM, F0_USA, B_USA, date_settlement, flag);

%% Plot of the surface of the implied volatilities

if flag == 1
    surface_vols(data_calib_EU,F0_EU);
    surface_vols(data_calib_USA,F0_USA);
end

%%

%% Multi-calibration comparison 1st round

% !WARNING:  calibration feasibility tests take a lot of computation time.
% The time used on a HP 64 bit, 16 GB RAM, Intel Core i7 was about 40
% minutes. This is due to the 33 consequent calibrations necessary: mean
% time for each calibration is 72 seconds.

% [params_removal, fun_eval, exit_condition] = multi_calib(data_calib_EU, data_calib_USA, F0_EU, B_EU, F0_USA, B_USA, date_settlement);

% Plots of the multicalib results 1st round

% if flag == 1
%     plot_multi_calib(dates_EU, dates_USA, fun_eval, 13, 20);
% end

%% Multi-calibration comparison 2nd round
% !WARNING:  calibration feasibility tests take a lot of computation time.
% The time used on a HP 64 bit, 16 GB RAM, Intel Core i7 was about 38
% minutes. This is due to the 32 consequent calibrations necessary: mean
% time for each calibration is 70 seconds.

% We remove the last date on the American side due to the 1st results
% [data_calib_USA, F0_USA, B_USA] = removal_expiry(data_calib_USA, F0_USA, B_USA, 20);
% 
% [params_removal, fun_eval, exit_condition] = multi_calib(data_calib_EU, data_calib_USA, F0_EU, B_EU, F0_USA, B_USA, date_settlement);

% Plots of the multicalib results 2nd round

% if flag == 1
%     plot_multi_calib(dates_EU, dates_USA, fun_eval, 13, 19);
% end

%% Multi-calibration comparison 3rd round
% !WARNING:  calibration feasibility tests take a lot of computation time.
% The time used on a HP 64 bit, 16 GB RAM, Intel Core i7 was about 36
% minutes. This is due to the 31 consequent calibrations necessary: mean
% time for each calibration is 68 seconds.

% We remove the last two dates on the American side due to the 2nd results
% [data_calib_USA, F0_USA, B_USA] = removal_expiry(data_calib_USA, F0_USA, B_USA, [19:20]);
% 
% [params_removal, fun_eval, exit_condition] = multi_calib(data_calib_EU, data_calib_USA, F0_EU, B_EU, F0_USA, B_USA, date_settlement);

% Plots of the multicalib results 2nd round

% if flag == 1
%     plot_multi_calib(dates_EU, dates_USA, fun_eval, 13, 18);
% end

%%

%% Removal of last american maturity
% Remove the last american maturity due to the appendix 3 considerations
% about the multi-calibration procedure
F = F0_USA; B = B_USA;

[data_calib_USA, F0_USA, B_USA] = removal_expiry(data_calib_USA, F0_USA, B_USA, 20);
[data_USA_OTM, ~] = removal_expiry(data_USA_OTM, F, B, 20);
[data_USA, ~] = removal_expiry(data_USA, F, B, 20);

clear F B;

%%

%% Calibration of the Xt model parameters

% Quantities of interest
% The initial parameters collected after various trials that minimizes the
% most the obj fun are the following ones; the structure is a bit specific
% but the fmincon is not able to find the global minimum otherwise
x0 = [0.3 -0.5 0.15 0.3 -0.5 0.15];
initial_cond = x0;

% Linear inequality constraints
A = [-1 0 0 0 0 0;
    0 0 -1 0 0 0;
    0 0 0 -1 0 0;
    0 0 0 0 0 -1];

b = [0; 0; 0; 0];

% Linear equality constraints
Aeq = []; beq = [];

% Lower and upper bounds
% Constraints of volvol on lb are required to stop a collapse of the 
% impvol structure if the volvol is too low
lb = [0.01; -Inf; 0; 0.01; -Inf; 0];
ub = [];

% Options for the visualization
options = optimset('MaxFunEvals', 3e3, 'Display', 'iter');

% Calibration
params_marginals = fmincon(@(params) calibration(params, data_calib_EU, data_calib_USA, ...
    F0_EU, B_EU, F0_USA, B_USA, date_settlement), x0, A, b, Aeq, beq, lb, ub, @(params) nonlinconstr(params), options);

% Display of the parameters on the console
params_USA = params_marginals(1:3);
params_EU = params_marginals(4:6);

% Explicit useful params
k1 = params_USA(1); k2 = params_EU(1);

%% Plots of the Levy model prices with calibrated values

if flag == 1

    % Plot over the entire curve
    plot_calls_puts_total(data_EU, F0_EU, B_EU, params_EU, date_settlement);
    plot_calls_puts_total(data_USA, F0_USA, B_USA, params_USA, date_settlement);

    % Plot over the filtered options
    plot_calls_puts(data_calib_EU, F0_EU, B_EU, params_EU, date_settlement);
    plot_calls_puts(data_calib_USA, F0_USA, B_USA, params_USA, date_settlement);

    % Plot the implied volatilities over the Calls
    plot_volatility_smiles(data_calib_EU, F0_EU, B_EU, params_EU, date_settlement)
    plot_volatility_smiles(data_calib_USA, F0_USA, B_USA, params_USA, date_settlement)

end

%%

%% POINT 7: Calibration of the historical correlation

% Initialization of the parameters
A = [-1 0 0; 0 -1 0; 0 0 -1]; b = [0; 0; 0];
Aeq = []; beq = [];
lb = zeros(1, 3); ub = [];

x0 = ones(1, 3);

params = fmincon(@(params) (sqrt(params(1) * params(2) / ((params(1) + params(3))*(params(2) + params(3)))) - rho_historical)^2, ...
    x0, A, b, Aeq, beq, lb, ub, @(params) nonlinconstr_corr(params, k1, k2), options);

% Extraction of the parameters
nu_USA = params(1); nu_EU = params(2);
nu_z = params(3);

% Console print of the calibration
disp_params(params_marginals, initial_cond, save_results);

%% Common and idiosynchratic parameters

% Compute the rho obtained from the model
rho_model_Levy = sqrt(params(1) * params(2) / ((params(1) + params(3))*(params(2) + params(3))));

error_rho = abs(rho_model_Levy - rho_historical);
disp('error for rho calibration:  ')
disp(error_rho);

% Compute the solution of the system
sol =  marginal_param(params_USA,params_EU , nu_z);

% Explicitate the parameters
a_USA = sol.x(1); a_EU = sol.x(2);
beta_z = sol.x(3); gamma_z = sol.x(4);

% Remaining parameters computation
beta_USA = params_USA(2) - a_USA * beta_z;                       % From condition (10.1)
gamma_USA = sqrt((params_USA(3)^2) - (a_USA^2) * (gamma_z^2));   % From condition (10.2)

beta_EU = params_EU(2) - a_EU * beta_z;
gamma_EU = sqrt(params_EU(3)^2 - a_EU ^ 2 * gamma_z^2);

% Idiosyncratic parameters USA
idiosync_USA = [nu_USA, beta_USA, gamma_USA, a_USA];

% Idiosyncratic parameters EU
idiosync_EU = [nu_EU, beta_EU, gamma_EU, a_EU];

% Systematic parameters
syst_Z = [nu_z , beta_z, gamma_z];

% Display of the parameters over the command window
disp_marginal_params(idiosync_USA , idiosync_EU , beta_z, gamma_z, nu_z, save_results);

%%

%% Point 8: Black model calibration

% Initialization of the parameters
x0 = 0.13;

% Linear inequality constraints
A = []; b = [];

% Linear equality constraints
Aeq = []; beq = [];

% Lower and upper bounds
lb = 0; ub = [];

% Options for the visualization
options = optimset('MaxFunEvals', 3e3, 'Display', 'iter');

% Calibration of sigma EU
sigma_EU = fmincon(@(sigma) blk_calibration(sigma, data_calib_EU, F0_EU, B_EU, date_settlement), ...
     x0, A, b, Aeq, beq, lb, ub, [], options);

% Calibration of sigma USA
sigma_USA = fmincon(@(sigma) blk_calibration(sigma, data_calib_USA, F0_USA, B_USA, date_settlement), ...
    x0, A, b, Aeq, beq, lb, ub, [], options);

%% Plots of the Black model prices with calibrated volatilities

if flag == 1
  
    % Plot over the entire curve
    blk_plot_calls_puts_total(data_EU, F0_EU, B_EU, sigma_EU, date_settlement);
    blk_plot_calls_puts_total(data_USA, F0_USA, B_USA, sigma_USA, date_settlement);

end


%% Point 9: Pricing of the certificate

% Computation of the pricing of the certificate
% The certificate is a derivative between the S&P500 and the EURO50 markets
% The payoff is the following:
% - If the EURO50 is below 95% of the initial value, the payoff is 0
% - Otherwise, the payoff is the difference between the S&P500 and the initial value

%% Point 9: Pricing of the certificate - LEVY

year_to_maturity = 1;

[rate_USA, TTM] = interp_pricing_params(datenum(data_calib_USA.datesExpiry), B_USA, date_settlement, year_to_maturity);
[rate_EU, ~] = interp_pricing_params(datenum(data_calib_EU.datesExpiry), B_EU, date_settlement, year_to_maturity);

S0_USA = data_USA.spot; S0_EU = data_EU.spot;

% Computation of the discount at 1y
B0_Levy = exp(-rate_USA * TTM);

S0_Levy = [S0_USA S0_EU];
rates_Levy = [rate_USA rate_EU];

[St_Levy, St_Levy_AV] = stock_simulation_Levy(idiosync_USA, idiosync_EU, syst_Z, params_USA, params_EU, S0_Levy, rates_Levy, TTM);

%% Normal Levy

% Unpacking the results
St_USA_Levy = St_Levy(:, 1);
St_EU_Levy = St_Levy(:, 2);

% Computation of the pricing certificate payoff 
indicator_Levy = St_EU_Levy < (0.95 * S0_EU);
certificate_payoff_Levy = max(St_USA_Levy - S0_USA, 0) .* indicator_Levy;

% Mean price and confidence interval
[mean_price_Levy, ~, IC_Levy] = normfit(B0_Levy * certificate_payoff_Levy);

%% Antithetic Levy

% Unpacking the results
St_USA_Levy_AV = St_Levy_AV(:, 1);
St_EU_Levy_AV = St_Levy_AV(:, 2);

% Computation of the pricing certificate payoff 
indicator_Levy_AV = St_EU_Levy_AV < (0.95 * S0_EU);
certificate_payoff_Levy_AV = max(St_USA_Levy_AV - S0_USA, 0) .* indicator_Levy_AV;

% Mean price and confidence interval
[mean_price_Levy_AV, ~, IC_Levy_AV] = normfit(B0_Levy * certificate_payoff_Levy_AV);

%% Levy semi-closed

price_Levy_closed = levy_semiclosed(S0_USA, rate_USA, rate_EU, params_USA, params_EU, idiosync_USA, idiosync_EU, syst_Z, TTM);

%%

%% Point 9: Pricing of the certificate - Brownian Motion

% set the time to maturity for the contract
year_to_maturity = 1;

% Computation of the rates and the time to maturity
[rate_USA, TTM] = interp_pricing_params(datenum(data_calib_USA.datesExpiry), B_USA, date_settlement, year_to_maturity);
[rate_EU, ~] = interp_pricing_params(datenum(data_calib_EU.datesExpiry), B_EU, date_settlement, year_to_maturity);

% Stock prices
S0_USA = data_USA.spot; S0_EU = data_EU.spot;

% Forward prices
F01_USA = S0_USA*exp(rate_USA*TTM);
F01_EU = S0_EU*exp(rate_EU*TTM);

% % Simulation of the underlying stock prices
[St_Black, St_Black_AV] = stock_simulation_Black([sigma_USA; sigma_EU], [F01_USA; F01_EU], ...
    rho_historical, TTM);

% Computation of the discount at 1y
B0_black = exp(-rate_USA * TTM);

%% NORMAL MC method

% Unpacking the results
St_USA_Black = St_Black(:, 1); St_EU_Black = St_Black(:, 2);

% Computation of the pricing certificate payoff 
indicator_Black = St_EU_Black < (0.95 * S0_EU);
certificate_payoff_Black = max(St_USA_Black - S0_USA, 0) .* indicator_Black;

% Mean price and confidence interval
[mean_price_Black, ~, IC_Black] = normfit(B0_black * certificate_payoff_Black);

%% ANTITHETIC MC method

% Unpacking the results
St_USA_Black_AV = St_Black_AV(:, 1); St_EU_Black_AV = St_Black_AV(:, 2);

% Computation of the pricing certificate payoff 
indicator_Black_AV = St_EU_Black_AV < (0.95 * S0_EU);
certificate_payoff_Black_AV = max(St_USA_Black_AV - S0_USA, 0) .* indicator_Black_AV;

% Mean price and confidence interval
[mean_price_Black_AV, ~, IC_Black_AV] = normfit(B0_black * certificate_payoff_Black_AV);

%% Closed formula

% Price computed via the closed formula
price_semiclosed = blk_semiclosed(data_USA.spot, rate_USA, rate_EU, sigma_USA, sigma_EU, rho_historical, TTM);

%% Display of the prices:

disp_contract_prices(mean_price_Levy,IC_Levy,mean_price_Levy_AV, IC_Levy_AV, ...
    mean_price_Black,IC_Black,mean_price_Black_AV,IC_Black_AV, ...
    price_Levy_closed, price_semiclosed);

%% End run time computation
elapsedTime = toc;
elapsedTimeMinutes = floor(elapsedTime / 60);
fprintf('Loop execution time: %.2f minutes and %.2f seconds \n', elapsedTimeMinutes , floor(elapsedTime-elapsedTimeMinutes*60));

File: Project2_MultivariatePricing/Python/Run_final_project_group2A.py
# Final FE Project: Multivariate Levy Pricing
# Group 2A

# Description:
# The project focus on the comparison of 2 different methods of model
# pricing; an exponential Levy framework with multiple parameters and a
# Black framework which takes consideration of the volatility only.
# Moreover, the project requires the pricing of an exotic derivative
# between the S&P500 and the EURO50 markets.

# Marco Maspes
# Andrea Tarditi
# Matteo Torba

# Libraries built-in
import numpy as np
import datetime as dt
import scipy.io as sio
import warnings

from scipy.optimize import minimize, fsolve
from datetime import datetime, timedelta
from scipy.interpolate import interp1d

# Libraries customized
import Dataset as obj_data
import calibration_functions as calib_fun
import forward_functions as fwd_fun
import pricing_functions as price_fun

# Disable all warnings
warnings.filterwarnings('ignore')

# Start run time computation
import time
start_time = time.time()

# Fix random seed for the simulation
np.random.seed(42)

#####################################################################################
####################           START THE CODE CREATION          #####################
#####################################################################################

# load files
date_settlement = dt.datetime(2023, 7, 9)

spxsx5e_returns = sio.loadmat('SPXSX5Ereturns.mat')

# Settings of initial datasets
data = sio.loadmat("OptionData.mat")
data_USA = obj_data.Dataset(data['mkt'])
data_EU = obj_data.Dataset(data['mkt_EU'])

# Removal of the last American object
data_USA._remove_last_element()

# Extract the components from the 'Returns' key
daily_returns = spxsx5e_returns['Returns'][0, 0]['Daily']
weekly_returns = spxsx5e_returns['Returns'][0, 0]['Weekly']
monthly_returns = spxsx5e_returns['Returns'][0, 0]['Monthly']
quarterly_returns = spxsx5e_returns['Returns'][0, 0]['Quarterly']
semiannually_returns = spxsx5e_returns['Returns'][0, 0]['SemiAnnually']
annually_returns = spxsx5e_returns['Returns'][0, 0]['Annually']

# Completing the EU and USA dataset with dates

# EU expiry dates
expiry_dates_EU = [
    '2023-07-21', '2023-08-18', '2023-09-15', '2023-10-20',
    '2023-11-17', '2023-12-15', '2024-01-19', '2024-02-16',
    '2024-03-15', '2024-06-21', '2024-09-20', '2024-12-20',
    '2025-06-20'
]

expiry_dates_EU = [dt.datetime.strptime(date, '%Y-%m-%d') for date in expiry_dates_EU]

data_EU._list_dates = expiry_dates_EU

# USA expiry dates
expiry_dates_USA = [
    '2023-07-21', '2023-08-18', '2023-09-15', '2023-10-20', '2023-11-17', '2023-12-15',
    '2024-01-19', '2024-02-16', '2024-03-15', '2024-04-19', '2024-05-17', '2024-06-21',
    '2024-07-19', '2024-09-20', '2024-12-20', '2025-06-20', '2025-12-19', '2026-12-18',
    '2027-12-17']

expiry_dates_USA = [dt.datetime.strptime(date, '%Y-%m-%d') for date in expiry_dates_USA]

data_USA._list_dates = expiry_dates_USA

# Year frac conventions
conv_ACT360 = 2
conv_ACT365 = 3
conv_30360_EU = 6

# Compute the historical correlation for daily returns
daily_correlation = calib_fun.hist_corr(annually_returns)

#%% #################################################################################
####################           FORWARD COMPUTATION              #####################
#####################################################################################

# Computation of the forward prices
F0_EU, B_EU = fwd_fun.forward_prices(data_EU)
F0_USA, B_USA = fwd_fun. forward_prices(data_USA)

#%% #################################################################################
####################           DATA REFINEMENT                  #####################
#####################################################################################

# Options selection and calibration
data_EU_OTM = calib_fun.OTM_preprocessing(data_EU, F0_EU)
data_USA_OTM = calib_fun.OTM_preprocessing(data_USA, F0_USA)

data_calib_EU = calib_fun.dataset_preprocessing(data_EU_OTM, F0_EU, B_EU, date_settlement)
data_calib_USA = calib_fun.dataset_preprocessing(data_USA_OTM, F0_USA, B_USA, date_settlement)

#%% #################################################################################
####################           CALIBRATION MARGINALS            #####################
#####################################################################################

# Calibration of the model
x0 = np.array([0.3, -0.5, 0.15, 0.3, -0.5, 0.15])

# Lower and upper bounds for the parameters
lb = np.array([0.01, -np.inf, 0, 0.01, -np.inf, 0])
ub = [None] * len(lb)

# Options for the optimization algorithm
options = {'maxiter': 3000, 'disp': True}

# Combine the constraints
nonlinear_constraints = {
    'type': 'eq',
    'fun': lambda x: calib_fun.nonlinconstr(x)[1]
}

# The calibration function should be passed without calling it
def calibration_function(params):
    return calib_fun.calibration(params, data_calib_EU, data_calib_USA, F0_EU, B_EU, F0_USA, B_USA, date_settlement)

# Minimize the calibration function with constraints and bounds
res = minimize(calibration_function, x0, method='SLSQP', bounds=[(lb[i], ub[i]) for i in range(len(lb))],
               constraints=nonlinear_constraints, options=options)

params_marginals = res.x

# Display of the parameters on the console
params_USA = params_marginals[:3]
params_EU = params_marginals[3:]

# Explicit useful params
k1 = params_USA[0]
k2 = params_EU[0]

#%% #################################################################################
################              DISPLAY PARAMETERS                    #################
#####################################################################################

print('---------------------------------')
print('k_USA:')
print(params_USA[0])
print('theta_USA:')
print(params_USA[1])
print('sigma_USA:')
print(params_USA[2])
print('---------------------------------')
print('k_EU:')
print(params_EU[0])
print('theta_EU:')
print(params_EU[1])
print('sigma_EU:')
print(params_EU[2])


#%% #################################################################################
################         CALIBRATION IDIOSYNC & SYSTEMATIC            ###############
#####################################################################################

# Try to solve a system to find nu_z, nu_1 and nu_2
rho_theoretical_matlab = 0.2911

def equations(x):
    nu_1, nu_2, nu_z = x

    eq1 = k1 - (nu_1 * nu_z)/(nu_1 + nu_z)
    eq2 = k2 - (nu_2 * nu_z)/(nu_2 + nu_z)
    eq3 = np.sqrt(k1 * k2)/nu_z - rho_theoretical_matlab

    return [eq1, eq2, eq3]


x0 = 0.2 * np.ones(3)
sol = fsolve(equations, x0)

nu_USA, nu_EU, nu_z = sol

# Compute the solution of the system
sol = calib_fun.marginal_param(params_USA, params_EU, nu_z)

# Explicitate the parameters
a_USA = sol[0]
a_EU = sol[1]
beta_z = sol[2]
gamma_z = sol[3]

# Remaining parameters computation
beta_USA = params_USA[1] - a_USA * beta_z
gamma_USA = np.sqrt(params_USA[2]**2 - a_USA**2 * gamma_z**2)
beta_EU = params_EU[1] - a_EU * beta_z
gamma_EU = np.sqrt(params_EU[2]**2 - a_EU**2 * gamma_z**2)

# Idiosyncratic parameters USA
idiosync_USA = [nu_USA, beta_USA, gamma_USA, a_USA]

# Idiosyncratic parameters EU
idiosync_EU = [nu_EU, beta_EU, gamma_EU, a_EU]

# Systematic parameters
syst_Z = [nu_z , beta_z, gamma_z]


#%% #################################################################################
################              DISPLAY PARAMETERS                    #################
#####################################################################################

print('---------------------------------')
print('nu_USA:')
print(nu_USA)
print('nu_EU:')
print(nu_EU)
print('nu_z:')
print(nu_z)

print('---------------------------------')
print('beta_USA:')
print(beta_USA)
print('beta_EU:')
print(beta_EU)
print('beta_z:')
print(beta_z)

print('---------------------------------')
print('gamma_USA:')
print(gamma_USA)
print('gamma_EU:')
print(gamma_EU)
print('gamma_z:')
print(gamma_z)

print('---------------------------------')
print('a_USA:')
print(a_USA)
print('a_EU:')
print(a_EU)


#%% #################################################################################
################                  CALIBRATION BLACK                 #################
#####################################################################################

# Initialization of the parameters
x0 = 1e-4

# Calibration of sigma EU
sigma_EU = minimize(lambda sigma: calib_fun.blk_calibration(sigma, data_calib_EU, F0_EU, B_EU, date_settlement), x0,
                     bounds=[(0, 1)], method='SLSQP').x

# Calibration of sigma USA
sigma_USA = minimize(lambda sigma: calib_fun.blk_calibration(sigma, data_calib_USA, F0_USA, B_USA, date_settlement), x0,
                     bounds=[(0, 1)], method='SLSQP').x

#%% #################################################################################
################                  PRICING LEVY                      #################
#####################################################################################

# Calculate rates for USA and EU
rate_USA, TTM_USA = price_fun.interp_pricing_params(data_calib_USA._list_dates, B_USA, date_settlement, 1)
rate_EU, TTM_EU = price_fun.interp_pricing_params(data_calib_EU._list_dates, B_EU, date_settlement, 1)

# Extract spot prices
S0_USA = data_calib_USA._spot
S0_EU = data_calib_EU._spot

# Compute discount at 1 year
B0_Levy = np.exp(-rate_USA * TTM_USA)

# Concatenate spot prices
S0_Levy = np.array([S0_USA, S0_EU])

# Concatenate rates
rates_Levy = np.array([rate_USA, rate_EU])

# Perform stock simulation
St_Levy = price_fun.stock_simulation_Levy(idiosync_USA, idiosync_EU, syst_Z, params_USA, params_EU, S0_Levy, \
                                          rates_Levy, TTM_USA)

# Unpack simulation results
St_USA_Levy = St_Levy[:, 0]
St_EU_Levy = St_Levy[:, 1]

# Compute certificate payoff
indicator_Levy = St_EU_Levy < (0.95 * S0_EU)
certificate_payoff_Levy = np.maximum(St_USA_Levy - S0_USA, 0) * indicator_Levy

# Compute mean price and confidence interval
mean_price_Levy = np.mean(B0_Levy * certificate_payoff_Levy)
std_dev_Levy = np.std(B0_Levy * certificate_payoff_Levy, ddof=1)
n = len(certificate_payoff_Levy)
z = 1.96  # 95% confidence interval
margin_of_error = z * (std_dev_Levy / np.sqrt(n))
IC_Levy = (mean_price_Levy - margin_of_error, mean_price_Levy + margin_of_error)


#%% #################################################################################
################                  PRICING BLACK                     #################
#####################################################################################

# Forward prices
F01_USA = S0_USA * np.exp(rate_USA * TTM_USA)
F01_EU = S0_EU * np.exp(rate_EU * TTM_EU)

# Simulation of the underlying stock prices
St_Black = price_fun.stock_simulation_Black(np.array([sigma_USA, sigma_EU]), np.array([F01_USA, F01_EU]),
                                                np.array([rate_USA, rate_EU]), 0.801, TTM_USA)

# Unpack simulation results
St_USA_Black = St_Black[:, 0]
St_EU_Black = St_Black[:, 1]

# Compute certificate payoff
indicator_Black = St_EU_Black < (0.95 * S0_EU)
certificate_payoff_Black = np.maximum(St_USA_Black - S0_USA, 0) * indicator_Black

B0_black = np.exp(-rate_USA * TTM_USA)

# Compute mean price and confidence interval
mean_price_Black = np.mean(B0_black * certificate_payoff_Black)
std_dev_Black = np.std(B0_black * certificate_payoff_Black, ddof=1)
n = len(certificate_payoff_Black)
z = 1.96  # 95% confidence interval
margin_of_error = z * (std_dev_Black / np.sqrt(n))
IC_Black = (mean_price_Black - margin_of_error, mean_price_Black + margin_of_error)

#%% #################################################################################
################                  DISPLAY PRICES                    #################
#####################################################################################

print('---------------------------------')
print('Levy Model')
print('Mean price: ', mean_price_Levy)
print('Confidence interval: ', IC_Levy)
print('---------------------------------')
print('Black Model')
print('Mean price: ', mean_price_Black)
print('Confidence interval: ',IC_Black)

# End run time computation and display it
end_time = time.time()
print('---------------------------------')
print(f"Run time: {end_time - start_time} seconds")